csrc/preseg_passes/propagate_shardings.cpp

```diff
@@ -115,17 +115,27 @@ std::vector<TensorView*> getOutputsWithoutMesh(Expr* expr) {
 // tvs and hence, can be propagated.
 std::unordered_set<ParallelType> getParallelTypesToPropagate(
     std::vector<TensorView*> tvs) {
-  // Get the set of parallel types seen on the loop domain of the given tvs.
+  std::unordered_set<ParallelType> all_parallel_types;
+  // Since we propagate across exprs, either all tvs are fusion inputs or none
+  // are.
+  if (tvs.at(0)->isFusionInput()) {
+    all_parallel_types = {ParallelType::Stream};
+  } else {
+    all_parallel_types = deviceAndStreamParallelTypes();
+  }
+
+  // Collect any DID or stream parallel types seen on the loop domain of the
+  // given tvs. For DID, only non-reduction dimensions are considered.
   std::unordered_set<ParallelType> existing_parallel_types;
   for (auto tv : tvs) {
     for (auto id : tv->getLoopDomain()) {
-      if (!id->isReduction() && id->isDeviceDim()) {
+      if ((id->isDeviceDim() && !id->isReduction()) || id->isStream()) {
```

Comment:
```suggestion
      if (id->isStream() || (id->isDeviceDim() && !id->isReduction())) {
```
So readers can short-circuit earlier.

Reply:
Also, code-comment why DIDx is treated specially. Our current comment describes the decision but not the reason.